lib/dry/cli/utils/files/file_system.rb

```diff
@@ -0,0 +1,188 @@
+          #
+          # @since x.x.x
+          def initialize(file: File, file_utils: FileUtils)
+            @file = file
+            @file_utils = file_utils
+          end
+
+          # Updates modification time (mtime) and access time (atime) of file(s)
+          # in list.
+          #
+          # Files are created if they don’t exist.
+          #
+          # @see https://ruby-doc.org/stdlib/libdoc/fileutils/rdoc/FileUtils.html#method-c-touch
+          #
+          # @param path [String, Array<String>] the target path
+          #
+          # @since x.x.x
+          # @api private
+          def touch(path, **kwargs)
+            file_utils.touch(path, **kwargs)
+          end
+
+          # Copies file content from `source` to `destination`
+          #
+          # @see https://ruby-doc.org/stdlib/libdoc/fileutils/rdoc/FileUtils.html#method-c-cp
+          #
+          # @param source [String] the file(s) or directory to copy
+          # @param destination [String] the directory destination
+          #
+          # @since x.x.x
+          # @api private
+          def cp(source, destination, **kwargs)
+            file_utils.cp(source, destination, **kwargs)
+          end
+
+          # Creates a directory and all its parent directories.
+          #
+          # The argument `path` is intended to be a **directory** that you want to
+          # explicitly create.
+          #
+          # @see #mkdir_p
+          # @see https://ruby-doc.org/stdlib/libdoc/fileutils/rdoc/FileUtils.html#method-c-mkdir_p
+          #
+          # @param path [String] the directory to create
+          #
+          # @example
+          #   require "dry/cli/utils/files/file_system"
+          #
+          #   fs = Dry::CLI::Utils::Files::FileSystem.new
+          #   fs.mkdir("/usr/var/project")
+          #   # creates all the directory structure (/usr/var/project)
+          #
+          # @since x.x.x
+          # @api private
+          def mkdir(path, **kwargs)
+            file_utils.mkdir_p(path, **kwargs)
+          end
+
+          # Creates a directory and all its parent directories.
+          #
+          # The argument `path` is intended to be a **file**, where its
+          # directory ancestors will be implicitly created.
+          #
+          # @see #mkdir
+          # @see https://ruby-doc.org/stdlib/libdoc/fileutils/rdoc/FileUtils.html#method-c-mkdir
+          #
+          # @param path [String] the file that will be in the directories that
+          #                      this method creates
+          # @example
+          #   require "dry/cli/utils/files/file_system"
+          #
+          #   fs = Dry::CLI::Utils::Files::FileSystem.new
+          #   fs.mkdir("/usr/var/project/file.rb")
+          #   # creates all the directory structure (/usr/var/project)
+          #   # where file.rb will eventually live
+          #
+          # @since x.x.x
+          # @api private
+          def mkdir_p(path, **kwargs)
```

Comment:
How about creating `touch_p` or extend current `touch` functionality to create directory structure and then create file there? IMO it will be more clear than the difference between mkdir (which uses mkdir_p of FileUtils) and mkdir_p of FileSystem here

Reply:
@IvanShamatov it doesn't create a file. You pass the path that includes the file, but it ignores it.

```ruby
fs.mkdir("/usr/var/project/file.rb") # => creates /usr/var/project
```